src/dnf-plugins/product-id/product-id.c

```diff
@@ -0,0 +1,636 @@
+        // with that product.
+        writeRepoMap(productDb);
+
+        // We have to free memory allocated for all items of repoAndProductIds. This should also handle
+        // activeRepoAndProductIds since the pointers in that array are pointing to the same underlying
+        // values at repoAndProductIds.
+        for (guint i=0; i < repoAndProductIds->len; i++) {
+            RepoProductId *repoProductId = g_ptr_array_index(repoAndProductIds, i);
+            free(repoProductId);
+        }
+
+        freeProductDb(productDb);
+        g_ptr_array_unref(repos);
+        g_ptr_array_unref(enabledRepos);
+        g_ptr_array_unref(repoAndProductIds);
+        g_ptr_array_unref(activeRepoAndProductIds);
+    }
+
+    return 1;
+}
+
+void writeRepoMap(ProductDb *productDb) {
+    GError *err = NULL;
+    writeProductDb(productDb, &err);
+
+    if (err) {
+        error("Unable to write productdb to file: %s", PRODUCTDB_FILE);
+    }
+}
+
+/**
+ * Find the list of repos that are actually enabled
+ * @param repos all available repos
+ * @param enabledRepos the list of enabled repos
+ */
+void getEnabled(const GPtrArray *repos, GPtrArray *enabledRepos) {
+    for (guint i = 0; i < repos->len; i++) {
+        DnfRepo* repo = g_ptr_array_index(repos, i);
+        bool enabled = (dnf_repo_get_enabled(repo) & DNF_REPO_ENABLED_PACKAGES) > 0;
+        if (enabled) {
+            g_ptr_array_add(enabledRepos, repo);
+        }
+    }
+}
+
+/**
+ * Find the list of repos that provide packages that are actually installed.
+ * @param repos all available repos
+ * @param activeRepoAndProductIds the list of repos providing active
+ */
+void getActive(DnfContext *context, const GPtrArray *repoAndProductIds, GPtrArray *activeRepoAndProductIds) {
+    DnfSack *dnfSack = dnf_context_get_sack(context);
+
+    // Create special sack object only for quering current rpmdb to get fresh list
+    // of installed packages. Quering dnfSack would not include just installed RPM
+    // package(s) or it would still include just removed package(s).
+    DnfSack *rpmDbSack = dnf_sack_new();
+    if(rpmDbSack == NULL) {
+        error("Unable to create new sack object for quering rpmdb");
+        return;
+    }
+
+    GError *tmp_err = NULL;
+    gboolean ret;
+    ret = dnf_sack_setup(rpmDbSack, 0, &tmp_err);
+    if (ret == FALSE) {
+        printError("Unable to setup new sack object", tmp_err);
+    }
+
+    ret = dnf_sack_load_system_repo(rpmDbSack, NULL, 0, &tmp_err);
+    if (ret == FALSE) {
+        printError("Unable to load system repo to sack object", tmp_err);
+    }
+
+    HyQuery query = hy_query_create_flags(rpmDbSack, 0);
+    hy_query_filter(query, HY_PKG_NAME, HY_GLOB, "*");
+    hy_query_filter(query, HY_REPO_NAME, HY_EQ, HY_SYSTEM_REPO_NAME);
+
+    GPtrArray *packageList = hy_query_run(query);
+    GPtrArray *installedPackages = g_ptr_array_sized_new(packageList->len);
```

Comment:
All packages in packagelist are installed, therefore a creation of installedpackages and testing of each package in array if it is installed is not needed.

Reply:
You are right again! :-)